internal/testdata/gen/gen.go

```diff
@@ -0,0 +1,393 @@
+		}
+		template.ExtraExtensions = append(template.ExtraExtensions, preIssuerExtension)
+	}
+
+	derBytes, err := x509.CreateCertificate(rand.Reader, &template, rootCACert, privKey.Public(), rootPrivKey)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create intermediate certificate: %v", err)
+	}
+
+	cert, err := x509.ParseCertificate(derBytes)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse intermediate certificate: %v", err)
+	}
+
+	return cert, nil
+}
+
+const (
+	organizationalUnit = "TrustFabric"
+	locality           = "London"
+	state              = "London"
+)
+
+type chainGenerator struct {
+	intermediateCert  *x509.Certificate
+	intermediateKey   any
+	leafCertPublicKey any
+}
+
+// newChainGenerator creates the chainGenerator.
+func newChainGenerator(intermediateCert *x509.Certificate, intermediateKey, leafCertPublicKey any) *chainGenerator {
+	return &chainGenerator{
+		intermediateCert:  intermediateCert,
+		intermediateKey:   intermediateKey,
+		leafCertPublicKey: leafCertPublicKey,
+	}
+}
+
+// certificate generates a deterministic TLS certificate by using integer as the serial number.
+func (g *chainGenerator) certificate(serialNumber int64, preCert bool) (*x509.Certificate, error) {
+	notBefore := time.Date(2023, 1, 1, 0, 0, 0, 0, time.UTC)
+	notAfter := time.Date(2024, 1, 1, 0, 0, 0, 0, time.UTC)
+
+	template := x509.Certificate{
+		SerialNumber: big.NewInt(serialNumber),
+		Subject: pkix.Name{
+			CommonName:         commonName,
+			Organization:       []string{organization},
+			OrganizationalUnit: []string{organizationalUnit},
+			Locality:           []string{locality},
+			Province:           []string{state},
+			Country:            []string{country},
+		},
+		NotBefore:             notBefore,
+		NotAfter:              notAfter,
+		KeyUsage:              x509.KeyUsageKeyEncipherment | x509.KeyUsageDigitalSignature,
+		ExtKeyUsage:           []x509.ExtKeyUsage{x509.ExtKeyUsageServerAuth},
+		BasicConstraintsValid: true,
+		DNSNames:              []string{commonName},
+	}
+
+	ctPoison := pkix.Extension{
+		Id:       asn1.ObjectIdentifier{1, 3, 6, 1, 4, 1, 11129, 2, 4, 3},
+		Critical: true,
+		Value:    []byte{0x05, 0x00}, // ASN.1 NULL
+	}
+
+	if preCert {
+		template.ExtraExtensions = append(template.ExtraExtensions, ctPoison)
+	}
+
+	derBytes, err := x509.CreateCertificate(rand.Reader, &template, g.intermediateCert, g.leafCertPublicKey, g.intermediateKey)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create leaf certificate: %v", err)
+	}
+
+	// Parse the DER-encoded certificate.
+	cert, err := x509.ParseCertificate(derBytes)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse root certificate: %v", err)
+	}
+
+	return cert, nil
+}
+
+func saveECDSAPrivateKeyPEM(key *ecdsa.PrivateKey, filename string) error {
+	// Marshal the private key to SEC1 ASN.1 DER.
+	derBytes, err := x509.MarshalECPrivateKey(key)
+	if err != nil {
+		return fmt.Errorf("failed to marshal EC private key: %w", err)
```

Comment:
`%w` -> `%v` unless you need to have callers be able to distinguish between different error instances/types (throughout)